### ISSUE TITLE
Debug support for displaying types

### ISSUE THREAD

**Issue by weaversa** [GaloisInc/cryptol#831]
In addition to the trace and traceVal functions, what do you think about a traceType (or showType) debug function?

When writing a complicated function, it might be nice to see the type (as Cryptol understands it) at a particular point.
Like this:

```cryptol
WStep:
    {n}
    (fin n, n >= 3) =>
    ([128] -> [128]) -> [n][64] -> [64] -> [n][64]
WStep CIPHk ([A] # Rs) t = [A'] # Rs'
  where
    [MSB, LSB] = showTypeVal "[MSB,LSB] has type:" (split (CIPHk (A # head Rs)))
    A'         = MSB ^ t
    Rs'        = tail Rs # [LSB]
```

Imagine `showType` working like `trace`. Or, maybe `trace` and `traceVal` could be augmented to include the printing of types?

**Comment by robdockins** [comment]
That would be relatively easy, I think, although it would have all the weird evaluation order issues that `trace` already has.

Maybe it would be better to integrate it into the type system though, so you could write, say, `split (CIPHk (A # head Rs) : ?my_type` have have Cryptol issue a warning where it tells you what type it inferred for `?my_type`.  We already have the machinery necessary for wildcard types (`asdf : _`), so I think it wouldn't be hard.

In the meantime, a trick I sometimes use in Haskell that works here too is to intentionally cause a type error and see what the compiler thinks the types are at that point:

```
WStep:
    {n}
    (fin n, n >= 3) =>
    ([128] -> [128]) -> [n][64] -> [64] -> [n][64]
WStep CIPHk ([A] # Rs) t = [A'] # Rs'
  where
    [MSB, LSB] = split (CIPHk (A # head Rs)) : ()
    A'         = MSB ^ t
    Rs'        = tail Rs # [LSB]
```
Loading this results in:


```
[error] at wstep.cry:7:18--7:23:
  Type mismatch:
    Expected type: ()
    Inferred type: [?parts`963][?each`964]
  where
  ?parts`963 is type argument 'parts' of 'split' at wstep.cry:7:18--7:23
  ?each`964 is type argument 'each' of 'split' at wstep.cry:7:18--7:23
[error] at wstep.cry:7:18--7:50:
  Type mismatch:
    Expected type: [2][64]
    Inferred type: ()
```

Here you can see that the compiler believes the `split` expression itself has under-constrained sizes (which is why the inferred type in the first message has variables in it), but that the context forces it to have type `[2][64]`.

**Comment by weaversa** [comment]
You're right! It doesn't make any sense to have a value-based command to print types when types are all known before execution. I like the idea of being able to name the type - this could really help diagnose type error messages.

It would be nice to see types after partially assigning types. For example, being able to ask to see ?my_type when `` WStep`{n=10} ``?
